### ISSUE TITLE
problem with modloadmsg option, Lmod and multiple lines

### ISSUE THREAD

**Issue by pescobar** [easybuilders/easybuild-framework#1067]
I am using EB-1.15.2 and Lmod-5.7.4

This easyconfig works great:

```
name = 'picard'
version = "1.107"

homepage = 'http://sourceforge.net/projects/picard'
description = """A set of tools (in Java) for working with next 
 generation sequencing data in the BAM format."""

toolchain = {'name': 'dummy', 'version': 'dummy'}

source_urls = [SOURCEFORGE_SOURCE]
sources = ['%(name)s-tools-%(version)s.zip']

dependencies = [('Java', '1.7.0_21')]

modloadmsg = """To list all the available utilities execute ls \\$EBROOTPICARD"""

moduleclass = 'bio'
```

But if I add a new line to modloadmsg, like this:

```
modloadmsg = """To list all the available utilities execute ls \\$EBROOTPICARD
To execute an utility run java -jar \\$EBROOTPICARD/utility.jar"""
```

I get this error in easybuild in sanity_check step:

```
== sanity checking...
== FAILED: Installation ended unsuccessfully (build directory: /tmp/easybuild/picard/1.107/dummy-dummy): build failed (first 300 chars): EasyBuild crashed with an error (at easybuild/framework/easyblock.py:1612 in sanity_check_step): Sanity check failed: loading fake module failed: 'Changing environment as dictated by module failed: name \'false\' is not defined (stdout: os.environ[\'LMOD_DEFAULT_MODULEPATH\'] = \'/scratch/easybuild-
== Results of the build can be found in the log file /scratch/easybuild-vxtZHt/easybuild-picard-1.107-20141021.235140.dxnEm.log
ERROR: EasyBuild crashed with an error (at easybuild/main.py:112 in build_and_install_software): Build of /import/bc2/home/admin/pub/ebconfigs/picard-1.107.eb failed (err: "build failed (first 300 chars): EasyBuild crashed with an error (at easybuild/framework/easyblock.py:1612 in sanity_check_step): Sanity check failed: loading fake module failed: 'Changing environment as dictated by module failed: name \\'false\\' is not defined (stdout: os.environ[\\'LMOD_DEFAULT_MODULEPATH\\'] = \\'/scratch/easybuild-")
```

When running in debug mode, this is the generated modulefile:

```
if [ module-info mode load ] {
        puts stderr     "To list all the available utilities execute ls \$EBROOTPICARD
To execute an utility run java -jar \$EBROOTPICARD/utility.jar"
}
```

@boegel confirms that this works when using TCL environment modules.

any clue @rtmclay ?

**Comment by michaelmayer2** [comment]
Hi guys,

wondering if there is a regression now - I have issues with modloadmsgs again when building texinfo, for example. It writes the following part of the lua module as a result of the modloadmsgs:

`if mode() == "load" then
    io.stderr:write("
    
    WARNING: This texinfo module has (re)defined the value for the environment variable \$TEXMFHOME.
    If you use a custom texmf directory (such as ~/texmf) you should copy files found in the
    new \$TEXMFHOME to your custom directory and reset the value of \$TEXMFHOME to point to that space:
    	cp -r $TEXMFHOME/* /path/to/your/texmf
    	export TEXMFHOME=/path/to/your/texmf
    
    ")`

A module file containing this cannot be loaded by EasyBuild - error message is 

`== 2016-08-23 09:00:18,835 easyblock.py:1900 WARNING Sanity check: loading fake module failed: 'Changing environment as dictated by module failed: name \'false\' is not defined (stdout: os.environ["LMOD_DEFAULT_MODULEPATH"] = "/tmp/eb-aeCCMP/tmpox4Ccm/all:/pstore/apps/.testing/modules/all";\nos.environ["MODULEPATH"] = "/tmp/eb-aeCCMP/tmpox4Ccm/all:/pstore/apps/.testing/modules/all";\nfalse\n, stderr: Lmod has detected the following error: Unable to load module:\ntexinfo/6.0-GCC-4.8.4\n     /tmp/eb-aeCCMP/tmpox4Ccm/all/texinfo/6.0-GCC-4.8.4.lua : [string\n"help([[Texinfo is the official documentation format of the GNU ..."]:26:\n**unfinished string near** \'"\'\n\nWhile processing the following module(s):\n    Module fullname        Module Filename\n    ---------------        ---------------\n    texinfo/6.0-GCC-4.8.4  /tmp/eb-aeCCMP/tmpox4Ccm/all/texinfo/6.0-GCC-4.8.4.lua\n\n\n)'`

This was obtained with Lua 5.3.3, Lmod 6.5 and EasyBuild 2.8.2.

**Comment by boegel** [comment]
@michaelmayer2 it looks like one of the `$` characters isn't properly escaped, which throws off Lmod

This:

```
 cp -r $TEXMFHOME/* /path/to/your/texmf
```

should be 

```
 cp -r \\$TEXMFHOME/* /path/to/your/texmf
```

@michaelmayer2 Can you check whether that fixes your problem?

**Comment by boegel** [comment]
@michaelmayer2 The problem is not in the (lack of) escaping of `$`, but in the multiline string, see feedback from @rtmclay (the Lmod developer) in https://gist.github.com/boegel/96fbec3d5bd131403f4c99ee717c00f4.

I'll fix that in the EasyBuild framework by properly escaping newlines that may appear in `modloadmsg` for Lua module files...

**Comment by boegel** [comment]
@michaelmayer2 problem is fixed after the changes in https://github.com/hpcugent/easybuild-framework/pull/1885 are applied (only the change from `"%s"` to `[==[%s]==]` in `easybuild/tools/module_generator.py` is relevant in case you want to apply this as a hotfix)

**Comment by michaelmayer2** [comment]
@boegel that's awesome - I am deeply impressed by this quick turnaround ! Thanks !

**Comment by boegel** [comment]
https://github.com/hpcugent/easybuild-framework/pull/1885 is merged, but does require Lmod 5.8 or more recent (to support multiline print statements in Lua module files)
